### ISSUE TITLE
Bug with macros and inheritance

### ISSUE THREAD

**Issue by duraki** [crystal-lang/crystal#7462]
Stumbled upon this issue and tried debugging it for quite a few hours. Asked for help from senior members but to no avail.

Issue seems to be either with inheritance of abstract class or using struct/stack in initialisation arguments.

```crystal
# save to /tmp/spec.cr
# run in two modes:
# $ CLASSES=Bar crystal spec spec.cr
# $ crystal spec spec.cr
# => unexpected output given below each macro

CLASSES = {
  "Foo" => Foo,
  "Baz" => Baz,
}

CLASSES_EXTRAS = {
  "Bar" => Bar,
}

CLASSES.merge!(CLASSES_EXTRAS)

pp CLASSES

module ClassHandler
  struct Context
    property worker

    def initialize(@worker : Int32)
    end
  end
end

abstract class AbstractFoo
  def initialize(@stdout : String, @i : ClassHandler::Context)
    pp "AbstractFoo initialized"
  end

  abstract def hi
end

abstract class CustomAbstractFoo < AbstractFoo
  def initialize(@stdout : String, @i : ClassHandler::Context)
    pp "CustomAbstractFoo initialized"
  end

  abstract def hi
end

class Foo < AbstractFoo
  def initialize(stdout : String)
    super(stdout, ctx)
    pp "Foo initialized"
  end

  def ctx : ClassHandler::Context
    ClassHandler::Context.new(0)
  end

  def hi
    "hello"
  end
end

class Bar < AbstractFoo
  def initialize(stdout : String)
    super(stdout, ctx)
    pp "Bar initialized"
  end

  def ctx : ClassHandler::Context
    ClassHandler::Context.new(1)
  end

  def hi
    "howdy"
  end
end

class Baz < CustomAbstractFoo
  def initialize(stdout : String)
    super(stdout, ctx)
    pp "Baz initialized"
  end

  def ctx : ClassHandler::Context
    ClassHandler::Context.new(2)
  end

  def hi
    "ola"
  end
end

stdout = "test string"

{% if c = env("CLASSES") %} # => Tries to initializes base (abstract) class instead of given class
  pp "Initiating class instance: " + {{c}}
  pp CLASSES[{{c}}].new(stdout).hi
  exit
{% end %}
# output:
# $ /tmp  CLASSES=Bar crystal spec spec.cr
# Error in line 1: while requiring "./spec.cr"

# in spec.cr:86: expanding macro

# {% if c = env("CLASSES") %}
# ^

# in spec.cr:86: expanding macro

# {% if c = env("CLASSES") %}
# ^

# in macro 'macro_4662393360' /private/tmp/spec.cr:86, line 3:

#    1.
#    2.   pp "Initiating class instance: " + "Bar"
# >  3.   pp CLASSES["Bar"].new(stdout).hi
#    4.   exit

# wrong number of arguments for 'AbstractFoo#initialize' (given 1, expected 2)

{% for cls in CLASSES.values %} # => Bar never reached
  p "=> ", {{cls}}
  pp {{cls}}.new(stdout).hi
{% end %}
# output:
# $ /tmp  crystal spec spec.cr
# {"Foo" => Foo, "Baz" => Baz, "Bar" => Bar}
# "=> "
# Foo
# "AbstractFoo initialized"
# "Foo initialized"
# "hello"
# "=> "
# Baz
# "CustomAbstractFoo initialized"
# "Baz initialized"
# "ola"
```

---

### Expected output

* **macro with ENV** - Construct given env `CLASSES` and execute `hi` method.
* **macro each CLASSES.values** - Iterate through given const and initialize the object instance, followed by running `hi` method.

### Current output

* **See output below each macro in above snippet**

---

### Tested on:

```
MacOS Mojave, latest update

Crystal 0.27.2 (2019-02-05)
LLVM: 6.0.1
Default target: x86_64-apple-macosx
```

**Comment by asterite** [comment]
Hi!

Reduced:

```crystal
abstract class AbstractFoo
  def initialize(i : Int32)
  end
end

abstract class CustomAbstractFoo < AbstractFoo
  def initialize(i : Int32)
  end
end

class Foo < AbstractFoo
  def initialize
    super(1)
  end
end

class Bar < AbstractFoo
  def initialize
    super(1)
  end
end

(Foo | Bar).new
```

The compiler is confused because `Foo | Baz` resolves to `AbstractFoo` ([explanation](https://crystal-lang.org/reference/syntax_and_semantics/virtual_and_abstract_types.html)) and then it goes and check that there's an `initialize` with no arguments for all subtypes, but for some reason it fails to notice that `CustomAbstractFoo` is abstract.

Also, if you comment the definition of `CustomAbstractFoo` it compiles but segfaults at runtime.

Finally, what would happen if you had `(AbstractFoo | Foo).new`, that `(Foo | AbstractFoo)` would also resolve to `AbstractFoo` but there's no way you can instantiate an abstract class, which leads us to #3835

In short: for now, please don't "dynamically" instantiate classes. It's broken.

My advice:

```crystal
case class_name
when "Foo" then Foo.new(...)
when "Bar" then Bar.new(...)
end
```

**Comment by asterite** [comment]
I guess one way to solve this would be to allow the above to work, but fail at runtime if you try to instantiate an abstract class. You lose a bit of security knowing your code will compile and run fine at runtime, but it gives more flexibility.
